text/204-golang-bindings.md

```diff
@@ -443,34 +453,180 @@ wrappers, e.g.
+    GetLaunchType()             LaunchType
+}
+
+type BaseServiceProps struct {
+    // Flattened properties generated from extended interface (i.e. Base ServiceOptions)
+    Cluster                ICluster
+    DesiredCount           *int
+    ServiceName            *string
+    MaxHealthyPercent      *int
+    MinHealthyPercent      *int
+    HealthCheckGracePeriod *Duration
+    CloudMapOptions        CloudMapOptions
+    PropagateTags          PropagatedTagSource
+    EnableECSManagedTags   *bool
+    DeploymentController   DeploymentController
+    // New properties introduced by BaseServiceProps
+    LaunchType             LaunchType
+}
+
+type baseServiceProps struct {
+    // NOTE: This could be baseServiceOptions, when both structs are defined in
+    // the same package. This could offer a marginal performance improvement,
+    // and perhaps a slightly more compact memory layout... Opting out of it
+    // on the other hand (as is done below), ensures the code path is always the
+    // same, regardless of the package layout.
+    IBaseServiceOptions // Embedded supertype
+
+    launchType             LaunchType
+}
+
+// Don't have to re-implement parent methods, as these are promoted from the
+// anonymous embed.
+func (b *baseServiceProps) GetLaunchType() { return b.launchType }
+
+func NewBaseServiceProps(args BaseServiceProps) IBaseServiceProps {
+    return &baseServiceProps{
+        IBaseServiceOptions: NewBaseServiceOptions(BaseServiceOptions{
+            Cluster:                args.Cluster,
+            DesiredCount:           args.DesiredCount,
+            ServiceName:            args.ServiceName,
+            MaxHealthyPercent:      args.MaxHealthyPercent,
+            MinHealthyPercent:      args.MinHealthyPercent,
+            HealthCheckGracePeriod: args.HealthCheckGracePeriod,
+            CloudMapOptions:        args.CloudMapOptions,
+            PropagateTags:          args.PropagateTags,
+            EnableECSManagedTags:   args.EnableECSManagedTags,
+            DeploymentController:   args.DeploymentController,
+        }),
+        launchType: args.LaunchType,
+    }
+}
+```
+
+The advantage of this approach is that adding a new optional property to a super struct no longer results in a breaking change (the field will be
+absent from the subtype until it is re-generated against the new parent), as the embedding technique guarantees methods are promoted from the
+supertype. It also guarantees immutability of built instances, as the actual implementation is not exported.
+
+The inconvenient is that this adds relatively heavy boilerplate around use of structs, where the name of the struct is repeated twice in sequence.
+This could lead to some user confusion (at least until the suer becomes familiar with this idiom). A typical call would look like so:
+
+```go
+// Pretending one can instantiate the BaseService construct directly
+NewBaseService(scope, "ID", NewBaseServiceProps(BaseServiceProps{
```

Comment:
Not sure I fully understand why `NewBaseService` cannot just accept a `BaseServiceProps` struct directly? What's the use case for passing in `IBaseServiceProps`?

Reply:
This is either *Approach 1* or *Approach 2* you're referring to. Both have caveats that cause non-breaking changes in TS to become breaking in Go.

But as we discussed, I've introduced *Approach 4*, which does not have these problems and will likely be more go-ey. It simply rids of the interfaces altogether, and uses the structs strictly by-value.